lisa/tests/staging/schedutil.py

```diff
@@ -0,0 +1,271 @@
+
+    """
+
+    def __init__(self, res_dir, plat_info, cpu):
+        super().__init__(res_dir, plat_info)
+        self.cpu = cpu
+
+    @requires_events('cpu_idle', 'cpu_frequency', 'sched_wakeup')
+    def estimate_nrg(self):
+        return self.plat_info['nrg-model'].estimate_from_trace(self.trace).sum(axis=1)
+
+    def get_avg_slack(self, only_negative=False):
+        pa = PerfAnalysis.from_dir(self.res_dir)
+
+        def get_slack(series):
+            if only_negative:
+                series = series[series < 0]
+
+            if series.empty:
+                return 0
+            else:
+                # average negative slack across all activations
+                return series.mean()
+
+        return {
+            task: get_slack(pa.get_df(task)["Slack"])
+            for task in pa.tasks
+        }
+
+    @LoadTrackingAnalysis.df_cpus_signal.used_events
+    @requires_events('schedutil_em')
+    def df_ramp_boost(self):
+        """
+        Return a dataframe with schedutil-related signals, sampled at the
+        frequency decisions timestamps for the CPU this bundle was executed on.
+
+        .. note:: The computed columns only take into account the CPU the test
+            was executing on. It currently does not handle multi-task workloads.
+        """
+        trace = self.trace
+        cpu = self.cpu
+
+        # schedutil_df also has a 'util' column that would conflict
+        schedutil_df = trace.df_events('schedutil_em')[['cpu', 'cost_margin']]
+        schedutil_df['from_schedutil'] = True
+
+        df_list = [
+            schedutil_df,
+            trace.analysis.load_tracking.df_cpus_signal('util'),
+            trace.analysis.load_tracking.df_cpus_signal('util_est_enqueued'),
+        ]
+
+        df = df_merge(df_list, filter_columns={'cpu': cpu})
+        df['from_schedutil'].fillna(value=False, inplace=True)
+        df.ffill(inplace=True)
+        df.dropna(inplace=True)
+
+        # Reconstitute how schedutil sees signals by subsampling the
+        # "master" dataframe, so we can look at signals coming from other
+        # dataframes
+        df = df[df['from_schedutil'] == True]
+        df.drop(columns=['from_schedutil'], inplace=True)
+
+        # If there are some NaN at the beginning, just drop some data rather
+        # than using fake data
+        df.dropna(inplace=True)
+
+        boost_points = (
+            # util_est_enqueued is the same as last freq update
+            (df['util_est_enqueued'].diff() == 0) &
+
+            # util_avg is increasing
+            (df['util'].diff() >= 0) &
+
+            # util_avg > util_est_enqueued
+            (df['util'] > df['util_est_enqueued'])
+        )
+        df['boost_points'] = boost_points
+
+        df['expected_cost_margin'] = (df['util'] - df['util_est_enqueued']).where(
+            cond=boost_points,
+            other=0,
+        )
+
+        # cost_margin values range from 0 to 1024
+        ENERGY_SCALE = 1024
```

Comment:
Is that meant to be the same scale as `SCHED_CAPACITY_SCALE`, or is just a [happy little accident](https://www.youtube.com/watch?v=_Tq5vXk0wTk)?

Reply:
That's the scale used to represent "per-1024-age" of costs as defined in:
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/linux/energy_model.h?h=v5.3#n17

That's not homogeneous to a capacity